pkgs/unified_analytics/lib/src/analytics.dart

```diff
@@ -546,32 +546,45 @@ class AnalyticsImpl implements Analytics {
     final collectionEvent =
         Event.analyticsCollectionEnabled(status: reportingBool);
 
-    // Construct the body of the request to signal
-    // telemetry status toggling
-    //
-    // We use don't use the sendEvent method because it may
-    // be blocked by the [telemetryEnabled] getter
-    final body = generateRequestBody(
-      clientId: _clientId,
-      eventName: collectionEvent.eventName,
-      eventData: collectionEvent.eventData,
-      userProperty: userProperty,
-    );
+    // The body of the request that will be sent to GA4
+    final Map<String, Object?> body;
 
-    _logHandler.save(data: body);
-
-    // Conditional logic for clearing contents of persisted
-    // files (except for config file) on opt out
-    if (!reportingBool) {
-      _sessionHandler.sessionFile.writeAsStringSync('');
-      _logHandler.logFile.writeAsStringSync('');
-      _clientIdFile.writeAsStringSync('');
-    } else {
+    if (reportingBool) {
       // Recreate the session and client id file; no need to
       // recreate the log file since it will only receives events
       // to persist from events sent
       Initializer.createClientIdFile(clientFile: _clientIdFile);
       Initializer.createSessionFile(sessionFile: _sessionHandler.sessionFile);
+
+      // Reread the client ID string so an empty string is not being
+      // sent to GA4 since the persisted files are cleared when a user
+      // decides to opt out of telemetry collection
+      _clientId = _clientIdFile.readAsStringSync();
+
+      // We must construct the body at this point after we have read in the
+      // new client id string that was generated
+      body = generateRequestBody(
+        clientId: _clientId,
+        eventName: collectionEvent.eventName,
+        eventData: collectionEvent.eventData,
+        userProperty: userProperty,
```

Comment:
Why have two identical `generateRequestBody` calls? Can we just have one after the if-else construct?

Reply:
I suppose I can move the call to generate the body outside of the conditionals, but then I will need to introduce a variable to store the correct client id to send in the request. Let me rewrite that and lmk what you think

Reply:
ah nevermind, the reason I have the body defined in two separate places is because when a user is opting in, we are saving that event, but when a user is opting out, we don't persist that event in the log file in order to keep it empty

Reply:
> you can see how we are saving it in on line 573 when we invoke

Oh yeah, I failed to notice that.